apps/rule-manager/app/controllers/RulesController.scala

```diff
@@ -279,4 +279,24 @@ class RulesController(
       case Left(error) => BadRequest(s"Invalid request: $error")
     }
   }
+
+  def csvImport() = Action { implicit request =>
+    val rules = for {
+      formData <- request.body.asMultipartFormData.toRight("No form data found in request")
+      file <- formData.file("file").toRight("No file found in request")
+      tag = formData.dataParts.get("tag") match {
+        case Some(tag) => tag.head
```

Comment:
Non-blocking – would it be better to model the tag as an `Option[String]` here, to avoid modelling an absent tag as an empty string? IIUC passing `tagName: Option[String]` to `csvImport` allows us to avoid a redundant tag find operation in that method, as well as a slightly odd error message when an empty string is interpolated into `s"Tag $tagName not found"`.

Reply:
That's a good point, thanks. I've attempted to make the change [here](https://github.com/guardian/typerighter/pull/477/commits/dd46453e01df8320ee527abcb69c650c1858accd). My poor scala is showing a bit, are those nested `case` statements okay or reprehensible?

Reply:
Not at all! Nothing wrong with a nested `match`, but while I'm no expert there are a few changes we could make:

- where you find yourself matching on nested values and there's no intervening computation, you might be able to destructure instead – for example, 

```scala
case Some(maybeTagId) =>
  maybeTagId match {
    case Some(tagId) =>
```

can become

```scala
case Some(Some(tagId)) => // happy path
// if you want to match on the unhappy paths, which is really nice to have
case Some(None) => 
case None =>
```

In this case, though, because it's `Option` all the way down, we can flatten the whole thing out as a for-comp. (I don't know if there's another way of including the logs as side-effects on the unhappy path than `orElse`, it'd be nice to avoid having to explicitly pass back `None`!)

```scala
for {
  tagName <- maybeTagName
  tag <- Tags.findAll().find(_.name == tagName).orElse {
    log.error(s"Tag $tagName not found")
    None
  }
  tagId <- tag.id.orElse {
    log.error(s"Tag $tagName has no ID")
    None
  }
} yield {
  RuleTagDraft.batchInsert(ruleIds.map(RuleTagDraft(_, tagId)))
}
```

Reply:
Thanks for this explanation! I went with the `for` comprehension